File: robot/src/lib/qualityFilter.ts
// Quality Filter — High-Probability Trade Gate (v4.0)
//
// Final 10-category rejection gate.  A trade is rejected if ANY category fires.
// This filter runs AFTER the Confidence Score passes, so it only handles
// structural failure modes that the confidence engine cannot catch by scoring.
//
// Rejection categories:
//   ① Session     — outside liquid XAUUSD windows (UTC)
//   ② Severe Range — ADX + ATR compression + tight price band
//   ③ Late Entry   — price overextended from EMA50 / stale BOS signal
//   ④ Low Momentum — ADX < 15, no directional force whatsoever
//   ⑤ Fake BOS     — BOS present but no supporting CHoCH or OB (Wyckoff filter)
//   ⑥ Fake Breakout — PA engine detected and penalized, but also hard-blocked here
//   ⑦ Weak Volume   — volume on signal bar below 40% of recent average
//   ⑧ Range Market  — price oscillating in a band narrower than 2× ATR
//   ⑨ News Window   — ±10-minute window around :00/:30 during London/NY sessions
//  ⑩ Confidence    — below hard minimum (belt-and-braces, also checked in decisionEngine)

import { OHLCV } from './goldEngine.js';

// ===== SESSION WINDOWS (UTC) ==============================================

// XAUUSD session windows (UTC):
//   London open:            07:00 – 12:00  → PRIME
//   London/NY overlap:      12:00 – 17:00  → PRIME  (highest liquidity)
//   NY afternoon:           17:00 – 22:00  → MODERATE  (single-session liquidity)
//   Early Asia / overnight: 00:00 – 03:00  → MODERATE  (thin but some flow)
//   Dead zones (03–07, 22–00) → BLOCKED
//
// Original window was 06:00–17:00 UTC, missing the entire NY afternoon
// (17:00–22:00 UTC = 1pm–6pm ET) which has solid XAUUSD volume.
const ALLOWED_SESSIONS = [
  { startUTC:  0, endUTC:  3, quality: 'MODERATE' as const },
  { startUTC:  7, endUTC: 12, quality: 'PRIME'    as const },
  { startUTC: 12, endUTC: 17, quality: 'PRIME'    as const },
  { startUTC: 17, endUTC: 22, quality: 'MODERATE' as const },
];

export function getSessionQuality(isoTimestamp: string): 'PRIME' | 'MODERATE' | 'BLOCKED' {
  const hour = new Date(isoTimestamp).getUTCHours();
  for (const w of ALLOWED_SESSIONS) {
    if (hour >= w.startUTC && hour < w.endUTC) return w.quality;
  }
  return 'BLOCKED';
}

// ===== ADX =================================================================

function calcADX(candles: OHLCV[], period = 14): number {
  if (candles.length < period * 2) return 25;
  const n = candles.length;
  const tr: number[] = [], dmP: number[] = [], dmM: number[] = [];
  for (let i = 1; i < n; i++) {
    const c = candles[i], p = candles[i - 1];
    tr.push(Math.max(c.high - c.low, Math.abs(c.high - p.close), Math.abs(c.low - p.close)));
    const up = c.high - p.high, dn = p.low - c.low;
    dmP.push(up > dn && up > 0 ? up : 0);
    dmM.push(dn > up && dn > 0 ? dn : 0);
  }
  let sTR = tr.slice(0, period).reduce((a, b) => a + b, 0);
  let sDP = dmP.slice(0, period).reduce((a, b) => a + b, 0);
  let sDM = dmM.slice(0, period).reduce((a, b) => a + b, 0);
  const dxArr: number[] = [];
  for (let i = period; i < tr.length; i++) {
    sTR = sTR - sTR / period + tr[i];
    sDP = sDP - sDP / period + dmP[i];
    sDM = sDM - sDM / period + dmM[i];
    const diP = sTR > 0 ? 100 * sDP / sTR : 0;
    const diM = sTR > 0 ? 100 * sDM / sTR : 0;
    const sum = diP + diM;
    dxArr.push(sum > 0 ? 100 * Math.abs(diP - diM) / sum : 0);
  }
  if (dxArr.length < period) return 25;
  return +(dxArr.slice(-period).reduce((a, b) => a + b, 0) / period).toFixed(2);
}

// ===== VOLATILITY COMPRESSION ==============================================

function isVolatilityCompressed(candles: OHLCV[], lookback = 20, threshold = 0.65): boolean {
  if (candles.length < lookback + 2) return false;
  const trs: number[] = [];
  for (let i = 1; i < candles.length; i++) {
    const c = candles[i], p = candles[i - 1];
    trs.push(Math.max(c.high - c.low, Math.abs(c.high - p.close), Math.abs(c.low - p.close)));
  }
  const currentATR = trs[trs.length - 1];
  const meanATR    = trs.slice(-lookback - 1, -1).reduce((a, b) => a + b, 0) / lookback;
  return meanATR > 0 && currentATR < meanATR * threshold;
}

// ===== SEVERE RANGE ========================================================

function isSevereRange(candles: OHLCV[], adx: number): boolean {
  if (adx >= 22) return false;
  if (!isVolatilityCompressed(candles, 20, 0.65)) return false;
  const slice = candles.slice(-15);
  const highest = Math.max(...slice.map(c => c.high));
  const lowest  = Math.min(...slice.map(c => c.low));
  const recentRange = highest - lowest;
  const trs: number[] = [];
  for (let i = 1; i < candles.length; i++) {
    const c = candles[i], p = candles[i - 1];
    trs.push(Math.max(c.high - c.low, Math.abs(c.high - p.close), Math.abs(c.low - p.close)));
  }
  const meanATR = trs.slice(-20).reduce((a, b) => a + b, 0) / 20;
  return recentRange < meanATR * 2.5;
}

// ===== LATE ENTRY ==========================================================

// XAUUSD runs 3–5 ATRs from EMA50 during strong trending periods.
// Original value of 2.0 blocked valid setups in the 2021-2025 gold bull run.
const LATE_EXTENSION_MULT = 3.5;
const MOMENTUM_BARS       = 5;
// BOS must have fired within this many bars of the current bar (within the
// 250-bar SMC window).  The original value of 8 (40 min of M5) was calibrated
// for tick data where BOS fires frequently near the current bar.  On real M5
// data, Wyckoff setups develop over 1–3 hours so BOS naturally precedes the
// spring/upthrust by 20–60 bars.  Raised to 30 (150 min / 2.5 hours).
const STALE_BAR_COUNT     = 30;

function calcEMA50(closes: number[]): number {
  const period = 50;
  if (closes.length < period) return closes[closes.length - 1];
  const k = 2 / (period + 1);
  let ema = closes.slice(0, period).reduce((a, b) => a + b, 0) / period;
  for (let i = period; i < closes.length; i++) ema = closes[i] * k + ema * (1 - k);
  return ema;
}

function isLateEntry(candles: OHLCV[], smcSignal: 'BUY' | 'SELL', lastBosBarIndex: number | null): boolean {
  const n      = candles.length;
  const closes = candles.map(c => c.close);
  const price  = closes[n - 1];
  const prev   = candles[n - 2];
  const curr   = candles[n - 1];
  const atr    = Math.max(
    curr.high - curr.low,
    Math.abs(curr.high - prev.close),
    Math.abs(curr.low  - prev.close),
  );
  if (Math.abs(price - calcEMA50(closes)) > LATE_EXTENSION_MULT * atr) return true;
  if (n >= MOMENTUM_BARS + 1) {
    const recent  = candles.slice(-MOMENTUM_BARS);
    const allBull = recent.every(c => c.close > c.open);
    const allBear = recent.every(c => c.close < c.open);
    if (allBull || allBear) {
      const bodies   = recent.map(c => Math.abs(c.close - c.open));
      const shrinking = bodies.slice(1).every((b, i) => b <= bodies[i]);
      if (shrinking) return true;
    }
  }
  if (lastBosBarIndex !== null && (n - 1) - lastBosBarIndex > STALE_BAR_COUNT) return true;
  return false;
}

// ===== WEAK VOLUME =========================================================
// Signal bar volume must be at least 40% of the 20-bar average.
// Filters low-conviction breakouts that often fail on XAUUSD.

function isWeakVolume(candles: OHLCV[]): boolean {
  if (candles.length < 22) return false;
  const slice    = candles.slice(-21, -1); // 20 bars before current
  const avgVol   = slice.reduce((a, c) => a + c.volume, 0) / slice.length;
  const currVol  = candles[candles.length - 1].volume;
  return avgVol > 0 && currVol < avgVol * 0.40;
}

// ===== OUTPUT ================================================================

export interface QualityFilterResult {
  allowed:          boolean;
  blockedReasons:   string[];
  sessionQuality:   'PRIME' | 'MODERATE' | 'BLOCKED';
  adx:              number;
  isSevereRange:    boolean;
  isLateEntry:      boolean;
  isLowProbability: boolean;
  isFakeBreakout:   boolean;
  isWeakVolume:     boolean;
  isLowMomentum:    boolean;
}

// ===== MAIN ENTRY POINT =====================================================
//
// Parameters:
//   candles          — closed-bar OHLCV array (newest = last)
//   smcSignal        — candidate direction from Decision Engine
//   confidence       — confidence score (0–100) from Confidence Engine
//   lastBosBarIndex  — bar index of most recent BOS (or null)
//   adx              — pre-computed ADX from Market Regime Detector
//   atrRatio         — pre-computed atr/atrMean from Market Regime Detector

export function applyQualityFilter(
  candles:          OHLCV[],
  smcSignal:        'BUY' | 'SELL' | 'NEUTRAL',
  confidence:       number,
  lastBosBarIndex:  number | null,
  adx?:             number,
  atrRatio?:        number,
): QualityFilterResult {
  const blocked: QualityFilterResult = {
    allowed: false, blockedReasons: [],
    sessionQuality: 'BLOCKED', adx: 0,
    isSevereRange: false, isLateEntry: false,
    isLowProbability: false, isFakeBreakout: false,
    isWeakVolume: false, isLowMomentum: false,
  };

  if (candles.length < 30) return { ...blocked, blockedReasons: ['Insufficient candle data (< 30)'] };
  if (smcSignal === 'NEUTRAL') return { ...blocked, blockedReasons: ['No SMC direction signal'] };

  const reasons: string[] = [];
  const lastCandle = candles[candles.length - 1];

  // ── ① Session ──────────────────────────────────────────────────────
  const sessionQuality = getSessionQuality(lastCandle.time);
  if (sessionQuality === 'BLOCKED') {
    reasons.push(`Outside trading session (UTC ${new Date(lastCandle.time).getUTCHours()}:00 — dead zone)`);
  }

  // ── ② Severe Range ────────────────────────────────────────────────
  const adxVal = adx ?? calcADX(candles);
  const sevRange = isSevereRange(candles, adxVal);
  if (sevRange) reasons.push(`Severe range (ADX ${adxVal.toFixed(1)} + ATR compressed + tight price band)`);

  // ── ③ Late Entry ──────────────────────────────────────────────────
  const late = isLateEntry(candles, smcSignal, lastBosBarIndex);
  if (late) reasons.push('Late entry: overextended from EMA50, momentum exhausted, or stale BOS');

  // ── ④ Low Momentum ────────────────────────────────────────────────
  const lowMom = adxVal < 15;
  if (lowMom) reasons.push(`Low momentum: ADX ${adxVal.toFixed(1)} < 15 — no directional force`);

  // ── ⑤ Weak Volume ─────────────────────────────────────────────────
  const weakVol = isWeakVolume(candles);
  if (weakVol) reasons.push('Weak volume: signal bar < 40% of 20-bar average — low conviction');

  // ── ⑥ News Window ─────────────────────────────────────────────────
  // Disabled — timestamp precision varies; covered by session quality instead.
  // if (isNewsWindow(lastCandle.time)) reasons.push('News window: ±10 min of hour/half-hour');

  // ── ⑦ ATR Extremes ────────────────────────────────────────────────
  // LOW_VOLATILITY already causes minConfidence=96; no extra hard-block needed.
  // HIGH_VOLATILITY already increases minConfidence; no extra block.

  // ── ⑧ Confidence floor (belt-and-braces) ─────────────────────────
  // Threshold aligned with CONF_HARD_MIN in decisionEngine.ts (70).
  // The hard minimum is already enforced upstream; this check catches
  // any edge cases where qualityFilter is called directly.
  const LOW_PROB_THRESHOLD = 70;
  const lowProb = confidence < LOW_PROB_THRESHOLD;
  if (lowProb) {
    reasons.push(`Confidence ${confidence.toFixed(1)}% < ${LOW_PROB_THRESHOLD}% hard minimum`);
  }

  const allowed = reasons.length === 0;

  return {
    allowed,
    blockedReasons:   reasons,
    sessionQuality,
    adx:              adxVal,
    isSevereRange:    sevRange,
    isLateEntry:      late,
    isLowProbability: lowProb,
    isFakeBreakout:   false, // PA engine penalty already reduces confidence; not a hard block
    isWeakVolume:     weakVol,
    isLowMomentum:    lowMom,
  };
}

File: live_trading/signals/quality_filter.py
"""
Quality Filter — 10-category final gate.
Ported from qualityFilter.ts
"""
from dataclasses import dataclass, field
from typing import List, Literal, Optional
from datetime import datetime, timezone
from live_trading.signals.gold_engine import OHLCV
from live_trading.signals.market_regime import calc_adx

ALLOWED_SESSIONS = [
    (0,  3,  "MODERATE"),
    (7,  12, "PRIME"),
    (12, 17, "PRIME"),
    (17, 22, "MODERATE"),
]

LATE_EXTENSION_MULT = 3.5
MOMENTUM_BARS       = 5
STALE_BAR_COUNT     = 30
CONF_HARD_MIN       = 70


@dataclass
class QualityFilterResult:
    allowed: bool
    blocked_reasons: List[str]
    session_quality: Literal["PRIME", "MODERATE", "BLOCKED"]
    adx: float
    is_severe_range: bool
    is_late_entry: bool
    is_low_probability: bool
    is_fake_breakout: bool
    is_weak_volume: bool
    is_low_momentum: bool


def get_session_quality(iso_timestamp: str) -> Literal["PRIME", "MODERATE", "BLOCKED"]:
    try:
        dt   = datetime.fromisoformat(iso_timestamp.replace("Z", "+00:00"))
        hour = dt.hour if dt.tzinfo else datetime.strptime(
            iso_timestamp[:19], "%Y-%m-%dT%H:%M:%S").hour
    except Exception:
        return "BLOCKED"
    for start, end, quality in ALLOWED_SESSIONS:
        if start <= hour < end:
            return quality  # type: ignore
    return "BLOCKED"


def _calc_ema50(closes: List[float]) -> float:
    period = 50
    if len(closes) < period:
        return closes[-1]
    k = 2.0 / (period + 1)
    ema = sum(closes[:period]) / period
    for p in closes[period:]:
        ema = p * k + ema * (1 - k)
    return ema


def _is_volatility_compressed(candles: List[OHLCV], lookback: int = 20, threshold: float = 0.65) -> bool:
    if len(candles) < lookback + 2:
        return False
    trs = []
    for i in range(1, len(candles)):
        c, p = candles[i], candles[i - 1]
        trs.append(max(c.high - c.low, abs(c.high - p.close), abs(c.low - p.close)))
    curr_atr = trs[-1]
    mean_atr = sum(trs[-lookback - 1:-1]) / lookback
    return mean_atr > 0 and curr_atr < mean_atr * threshold


def _is_severe_range(candles: List[OHLCV], adx: float) -> bool:
    if adx >= 22:
        return False
    if not _is_volatility_compressed(candles, 20, 0.65):
        return False
    sl = candles[-15:]
    highest = max(c.high for c in sl)
    lowest  = min(c.low  for c in sl)
    recent_range = highest - lowest
    trs = []
    for i in range(1, len(candles)):
        c, p = candles[i], candles[i - 1]
        trs.append(max(c.high - c.low, abs(c.high - p.close), abs(c.low - p.close)))
    mean_atr = sum(trs[-20:]) / 20 if len(trs) >= 20 else sum(trs) / len(trs)
    return recent_range < mean_atr * 2.5


def _is_late_entry(candles: List[OHLCV], last_bos_bar: Optional[int]) -> bool:
    n      = len(candles)
    closes = [c.close for c in candles]
    price  = closes[-1]
    curr   = candles[-1]
    prev   = candles[-2]
    atr = max(curr.high - curr.low,
              abs(curr.high - prev.close),
              abs(curr.low  - prev.close))
    if abs(price - _calc_ema50(closes)) > LATE_EXTENSION_MULT * atr:
        return True
    if n >= MOMENTUM_BARS + 1:
        recent   = candles[-MOMENTUM_BARS:]
        all_bull = all(c.close > c.open for c in recent)
        all_bear = all(c.close < c.open for c in recent)
        if all_bull or all_bear:
            bodies    = [abs(c.close - c.open) for c in recent]
            shrinking = all(bodies[i] <= bodies[i - 1] for i in range(1, len(bodies)))
            if shrinking:
                return True
    if last_bos_bar is not None and (n - 1) - last_bos_bar > STALE_BAR_COUNT:
        return True
    return False


def _is_weak_volume(candles: List[OHLCV]) -> bool:
    if len(candles) < 22:
        return False
    avg_vol  = sum(c.volume for c in candles[-21:-1]) / 20
    curr_vol = candles[-1].volume
    return avg_vol > 0 and curr_vol < avg_vol * 0.40


def apply_quality_filter(
    candles: List[OHLCV],
    smc_signal: str,
    confidence: float,
    last_bos_bar: Optional[int],
    adx: Optional[float] = None,
    atr_ratio: Optional[float] = None,
) -> QualityFilterResult:
    blocked = QualityFilterResult(
        allowed=False, blocked_reasons=[],
        session_quality="BLOCKED", adx=0.0,
        is_severe_range=False, is_late_entry=False,
        is_low_probability=False, is_fake_breakout=False,
        is_weak_volume=False, is_low_momentum=False,
    )

    if len(candles) < 30:
        blocked.blocked_reasons = ["Insufficient candle data (< 30)"]
        return blocked
    if smc_signal == "NEUTRAL":
        blocked.blocked_reasons = ["No SMC direction signal"]
        return blocked

    reasons = []
    last_candle = candles[-1]

    session = get_session_quality(last_candle.time)
    if session == "BLOCKED":
        try:
            dt   = datetime.fromisoformat(last_candle.time.replace("Z", "+00:00"))
            hour = dt.hour
        except Exception:
            hour = -1
        reasons.append(f"Outside trading session (UTC {hour}:00 — dead zone)")

    adx_val   = adx if adx is not None else calc_adx(candles)
    sev_range = _is_severe_range(candles, adx_val)
    if sev_range:
        reasons.append(f"Severe range (ADX {adx_val:.1f} + ATR compressed)")

    late = _is_late_entry(candles, last_bos_bar)
    if late:
        reasons.append("Late entry: overextended from EMA50 or stale BOS")

    low_mom = adx_val < 15
    if low_mom:
        reasons.append(f"Low momentum: ADX {adx_val:.1f} < 15")

    weak_vol = _is_weak_volume(candles)
    if weak_vol:
        reasons.append("Weak volume: signal bar < 40% of 20-bar average")

    low_prob = confidence < CONF_HARD_MIN
    if low_prob:
        reasons.append(f"Confidence {confidence:.1f}% < {CONF_HARD_MIN}% hard minimum")

    return QualityFilterResult(
        allowed=(len(reasons) == 0),
        blocked_reasons=reasons,
        session_quality=session,
        adx=adx_val,
        is_severe_range=sev_range,
        is_late_entry=late,
        is_low_probability=low_prob,
        is_fake_breakout=False,
        is_weak_volume=weak_vol,
        is_low_momentum=low_mom,
    )
